pkg/couchdb/couchdb.go

```diff
@@ -754,6 +753,56 @@ func GetAllDocs(db Database, doctype string, req *AllDocsRequest, results interf
 	return json.Unmarshal(data, results)
 }
 
+// ForeachDocs traverse all the documents from the given database with the
+// specified doctype and calls a function for each document.
+func ForeachDocs(db Database, doctype string, doc interface{}, fn func() error) error {
+	var startKey string
+	limit := 100
+	for {
+		skip := 0
+		if startKey != "" {
+			skip = 1
+		}
+		req := &AllDocsRequest{
+			StartKeyDocID: startKey,
+			Skip:          skip,
+			Limit:         limit,
+		}
+		v, err := query.Values(req)
+		if err != nil {
+			return err
+		}
+		v.Add("include_docs", "true")
+
+		var res AllDocsResponse
+		url := makeDBName(db, doctype) + "/_all_docs?" + v.Encode()
+		err = makeRequest(db, "GET", url, nil, &res)
+		if err != nil {
+			return err
+		}
+
+		var count int
+		startKey = ""
+		for _, row := range res.Rows {
+			if !strings.HasPrefix(row.ID, "_design") {
+				if err = json.Unmarshal(row.Doc, doc); err != nil {
```

Comment:
Do you think it is safe to reuse the same document? For example, if one row has a value for on optional field, and the next row hasn't it, won't the doc will keep it?